portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,45 +37,102 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private final ArrayList<UserComment> userComments = new ArrayList<>();
+  private final String defaultMaxComment = "20";
 
   /*
    * Called when a client submits a GET request to the /data URL
    * Displays all recorded user comments on page
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int maxCommentDisplay = Integer.parseInt(getFieldFromResponse(request, "maxcomments", defaultMaxComment));
+
+    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+    Query<Entity> query =
+        Query.newEntityQueryBuilder()
+          .setKind("Comment")
+          .setOrderBy(StructuredQuery.OrderBy.desc("time"))
+          .setLimit(maxCommentDisplay)
+          .build();
+    QueryResults<Entity> results = datastore.run(query);
+
+    ArrayList<UserComment> comments = new ArrayList<>();
+    while(results.hasNext()) {
+      Entity entity = results.next();
+      long id = entity.getKey().getId();
+      String name = entity.getString("name");
+      String email = entity.getString("email");
+      String time =  String.valueOf(entity.getLong("time"));
```

Comment:
Do we need to convert this to string?

If we directly put long in the json response, the client could also skip parseInt.
This might be problematic because Java and JS have different precisions for long integers. Java supports 64-bit integer but JavaScript only supports 53-bit, but I guess we don't need to worry about that until June 5, 2255 :)
Also the parseInt still has the same issue for precisions, so if you want your code to be correct after 2255 you could use BigInt :) 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt

Reply:
Hmm I don't think this website will be as wildly popular then :)

Reply:
Right, then I suggest to pass long instead of string and avoid parsings.

```
Long time = entity.getLong("time");
```

and change `@AutoValue` to use Long type instead of String. (Note that you might need to use `Long` instead of `long`, since AutoValue assumes everything is an Object instead of a primitive type.

And in the client, you do n't need to call `parseInt`.